{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Lab 04 Files and Modules\n",
    "\n",
    "# Deliverables (50 pts total):\n",
    " - sequenceAnalysis.py module - 31 total points\n",
    " - classes included: \n",
    "     - NucParams, \n",
    "     - ProteinParams, \n",
    "     - FastAreader\n",
    " - genomeAnalyzer.py - 15 total points\n",
    " - Lab 4 notebook with Inspection Intro Cell and Inspection Results Cell completed( 4 pts)\n",
    " - possible extra credit - 10 additional points\n",
    " - Due: Monday May 8, 2023 11:55pm\n",
    "\n",
    "\n",
    "Congratulations. You have started to build an inventory of some pretty useful functions.  Because these are written as classes, you can easily reuse them. Your ProteinParam class is your first deposit into your very own sequenceAnalysis toolkit.  In Python, these toolkits are called modules.\n",
    "\n",
    "We are also going to start using real sequence files.  The fastA format, described here: en.wikipedia.org/wiki/FASTA_format is very convenient to use and fully capable of storing sequences of many types. You will be reading these from an input file for this assignment.\n",
    "\n",
    "## Genomic analysis\n",
    "\n",
    "There are a few things that we can do that mirror and extend the analyses that we did previously on protein sequences. We can calculate composition statistics on the genome (gc content for example), we can calculate relative codon usage in the genome, and we can calculate amino acid composition by translating those codons used the genome.\n",
    "\n",
    "For this lab, I have provided a NucParams class, with the required methods that it implements (see below). You will need to design and write those methods, and these are to be placed in a file called sequenceAnalysis.py This is a __*module*__ that you can use from now on.\n",
    "\n",
    "You will also need to place the ProteinParams class from Lab 3 into this module. This class will not be used for this assignment, but place it into your toolbox.\n",
    "\n",
    "I have written the FastAreader class.  It is included below. Keep it as is part of your module for now, you may decide to keep it somewhere else later.\n",
    "\n",
    "The input file for this assignment will be named testGenome.fa, and is available in Canvas. You will not need to submit testGenome.fa, but it will be necessary for your testing.  For development and testing, create a new directory (Lab04) and place the data file (testGenome.fa), your Lab04 notebook, your program (genomeAnalyzer.py), your new module (sequenceAnalysis.py).\n",
    "\n",
    "\n",
    "## Hints\n",
    "\n",
    " - Python modules have the .py extension as files, but when they are imported, the name without the extension is used in the import statement in your program.\n",
    "\n",
    " - File placement: Make sure to place your notebook, program, sequenceAnalysis module and the required data files in the same folder. This will allow Python to find them. Read over the FastAreader usage to see how to specify file names that you can use for your data.\n",
    "\n",
    "## Codon frequency calculations\n",
    "\n",
    "Notice that NucParams does all of the counting you need. It is responsible for counts of codons and their translated amino acids.\n",
    "\n",
    "Your genomeAnalyzer.py program has the task of determining which codons are preferred for each of the amino acids and calculating the relative percentage.  For any given amino acid, the relative codon usage (percentages) should sum to 100.0%. Notice that Methionine and Tryptophan only have 1 codon that codes for those, so these will have relative codon usages of 100%.\n",
    "\n",
    "For example: Lysine is coded by both AAA (607) and AAG (917) (example counts in parentheses).  From our aaComposition() method, we are given the aaComposition dictionary and we can lookup 'K' to find 1524 counts (these came from those 607+917 codons).  We can then calculate 607/1524 for AAA and 917/1524 for AAG.  The associated percentages are thus: 39.8 for AAA and 60.2 for AAG.\n",
    "\n",
    "AAA = 607/1524 * 100 = 39.8%\n",
    "\n",
    "AAG = 917/1524 * 100 = 60.2%\n",
    "\n",
    "\n",
    "## Design specification - sequenceAnalysis.py\n",
    "\n",
    "### NucParams class\n",
    "\n",
    "#### \\_\\_init\\_\\_\n",
    "\n",
    "The constructor of the class has one optional parameter, a sequence of type string. It may include upper or lower case letters of the set {ACGTUN} or whitespace.  These will be gene sequences and they begin in frame 1.  In other words the first 3 letters of the sequence encode the first AA of the sequence. Carefully consider in what form this class should maintain its data. Is a string the best structure? This class (NucParams) is intended to be very similar to ProteinParam. Make sure to read addSequence() before making this decision, and remember that objects of this class may need to handle an arbitrarily large number of sequences (hint:  dictionaries are good). As a second hint, notice that __init__ and addSequence are doing VERY similar things - you could just make one of them do most of the work.\n",
    "\n",
    "#### addSequence() - 5 pts\n",
    "\n",
    "This method must accept new sequences, from the {ACGTUN} alphabet, and each sequence can be presumed to start in frame 1. This data must be added to the data that you were given with the __init__ method (if any).\n",
    "\n",
    "#### aaComposition() - 6 pts\n",
    "\n",
    "This method will return a dictionary of counts over the 20 amino acids and stop codons.  This dictionary is VERY similar to the lab 3 aaComposition, though you must decode the codon first. The translation table from codon to AA is provided. You are counting amino acids by translating from the proper codon table.\n",
    "\n",
    "#### nucComposition() - 10 pts\n",
    "\n",
    "This method returns a dictionary of counts of valid nucleotides found in the analysis. (ACGTNU}. If you were given RNA nucleotides, they should be counted as RNA nucleotides. If you were given DNA nucleotides, they should be counted as DNA nucleotides. Any N bases found should be counted also. Invalid bases are to be ignored in this dictionary.\n",
    "\n",
    "#### codonComposition() - 10 pts\n",
    "\n",
    "This dictionary returns counts of codons. Presume that sequences start in frame 1, accept the alphabet {ACGTUN} and store codons in RNA format, along with their counts. __Any codons found with invalid bases should be discarded__. Discard codons that contain N bases. This means that all codon counts are stored as RNA codons, even if the input happens to be DNA. If you discard a codon, make sure to not alter the frame of subsequent codons.\n",
    "\n",
    "#### nucCount()\n",
    "\n",
    "This returns an integer value, summing every valid nucleotide {ACGTUN} found.  This value should exactly equal the sum over the nucleotide composition dictionary."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Design specification - genomeAnalyzer.py\n",
    "\n",
    "This program must import your sequenceAnalysis module.\n",
    "It is responsible for preparing the summaries and final display of the data.\n",
    "\n",
    "## Input must be from STDIN\n",
    "Your FastaReader object will read data from sys.stdin if it is not given a filename. Notice that the filename is specified as a parameter to your main() function. See the last line of the cell containing main() in this notebook:\n",
    " - main( 'testGenome.fa').\n",
    "\n",
    "When you move main() to genomeAnalyzer.py, you can delete the parameter:\n",
    " - main()\n",
    " this will cause the filename parameter to be *None* (default parameter) and your FastaReader object will interpret a filename of *None* as a request to use sys.stdin. ( see the doOpen() method for further info).\n",
    "\n",
    "Your notebook will be used for inspections. For the notebook, you will need to specify the input filename. For genomeAnalyzer.py, you will need to delete the parameter.\n",
    "\n",
    "\n",
    "### Output format - 15 pts\n",
    "\n",
    "The function to output the results of your analysis has specific formatting rules that you must follow to get full credit. These rules are as follows:\n",
    "\n",
    " - First line: sequence length = X.XX Mb with two digits after the decimal and labeled Mb (you need to calculate the number of bases in Mb).\n",
    " - second line is blank\n",
    " - third line: GC content = XX.X% as a percentage with one digit after the decimal\n",
    " - fourth line is blank\n",
    " - lines 5 - 68 are the output statistics on relative codon usage for each codon ordered by codon within each amino acid group as follows:"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "XXX : A F (D)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "where XXX is the three letters for an RNA codon, A is the 1-letter amino acid code, F is relative codon frequency, use {:5.1f} for the format, and D is for codon count, use the format {:6d}. There is a single space between each of these fields.\n",
    "For example ( this is not representative of any real genome ):"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sequence length = 3.14 Mb  # 2.21 in given file\n",
    "\n",
    "GC content = 60.2%\n",
    "\n",
    "UAA : -  32.6 (  1041)\n",
    "UAG : -  38.6 (  1230)\n",
    "UGA : -  28.8 (   918)\n",
    "GCA : A  14.1 ( 10605)\n",
    "GCC : A  40.5 ( 30524)\n",
    "GCG : A  30.5 ( 22991)\n",
    "GCU : A  14.9 ( 11238)\n",
    "UGC : C  67.2 (  4653)\n",
    "UGU : C  32.8 (  2270)\n",
    "\n",
    "..."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output). \n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - contain proper docstrings for the program, classes, modules and any public functions.\n",
    " - Contain in-line comments\n",
    " \n",
    "## Extra credit - 10 pts possible\n",
    "\n",
    "You now have a very powerful set of classes for evaluating genomes. Write a compareGenomes.py program that compares GC content, aaComposition and relative codon bias of 2 genomes. You will have a halophile genome and a hyperthermophile genome to compare.\n",
    "\n",
    "Submit your code using canvas\n",
    "\n",
    "Congratulations, you have finished your fourth lab assignment!"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "nteract": {
     "transient": {
      "deleting": false
     }
    }
   },
   "source": [
    "## Inspection Intro\n",
    "Provide design level information for your inspection team here. How do you input data to avoid having to read every sequence from the genome into memory? Where are your composition dictionaries initialized? How does data get added to those composition dictionaries?"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Nuc Params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "class NucParams:\n",
    "    rnaCodonTable = {\n",
    "    # RNA codon table\n",
    "    # U\n",
    "    'UUU': 'F', 'UCU': 'S', 'UAU': 'Y', 'UGU': 'C',  # UxU\n",
    "    'UUC': 'F', 'UCC': 'S', 'UAC': 'Y', 'UGC': 'C',  # UxC\n",
    "    'UUA': 'L', 'UCA': 'S', 'UAA': '-', 'UGA': '-',  # UxA\n",
    "    'UUG': 'L', 'UCG': 'S', 'UAG': '-', 'UGG': 'W',  # UxG\n",
    "    # C\n",
    "    'CUU': 'L', 'CCU': 'P', 'CAU': 'H', 'CGU': 'R',  # CxU\n",
    "    'CUC': 'L', 'CCC': 'P', 'CAC': 'H', 'CGC': 'R',  # CxC\n",
    "    'CUA': 'L', 'CCA': 'P', 'CAA': 'Q', 'CGA': 'R',  # CxA\n",
    "    'CUG': 'L', 'CCG': 'P', 'CAG': 'Q', 'CGG': 'R',  # CxG\n",
    "    # A\n",
    "    'AUU': 'I', 'ACU': 'T', 'AAU': 'N', 'AGU': 'S',  # AxU\n",
    "    'AUC': 'I', 'ACC': 'T', 'AAC': 'N', 'AGC': 'S',  # AxC\n",
    "    'AUA': 'I', 'ACA': 'T', 'AAA': 'K', 'AGA': 'R',  # AxA\n",
    "    'AUG': 'M', 'ACG': 'T', 'AAG': 'K', 'AGG': 'R',  # AxG\n",
    "    # G\n",
    "    'GUU': 'V', 'GCU': 'A', 'GAU': 'D', 'GGU': 'G',  # GxU\n",
    "    'GUC': 'V', 'GCC': 'A', 'GAC': 'D', 'GGC': 'G',  # GxC\n",
    "    'GUA': 'V', 'GCA': 'A', 'GAA': 'E', 'GGA': 'G',  # GxA\n",
    "    'GUG': 'V', 'GCG': 'A', 'GAG': 'E', 'GGG': 'G'  # GxG\n",
    "    }\n",
    "    dnaCodonTable = {key.replace('U','T'):value for key, value in rnaCodonTable.items()}\n",
    "\n",
    "    def __init__ (self, inString=''):\n",
    "        pass\n",
    "        \n",
    "    def addSequence (self, inSeq):\n",
    "        pass\n",
    "    def aaComposition(self):\n",
    "        return self.aaComp\n",
    "    def nucComposition(self):\n",
    "        return self.nucComp\n",
    "    def codonComposition(self):\n",
    "        return self.codonComp\n",
    "    def nucCount(self):\n",
    "        return sum(self.nucComp.values())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## FastAreader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "class FastAreader :\n",
    "    ''' \n",
    "    Define objects to read FastA files.\n",
    "    \n",
    "    instantiation: \n",
    "    thisReader = FastAreader ('testTiny.fa')\n",
    "    usage:\n",
    "    for head, seq in thisReader.readFasta():\n",
    "        print (head,seq)\n",
    "    '''\n",
    "    def __init__ (self, fname=None):\n",
    "        '''contructor: saves attribute fname '''\n",
    "        self.fname = fname\n",
    "            \n",
    "    def doOpen (self):\n",
    "        ''' Handle file opens, allowing STDIN.'''\n",
    "        if self.fname is None:\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "        \n",
    "    def readFasta (self):\n",
    "        ''' Read an entire FastA record and return the sequence header/sequence'''\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        \n",
    "        with self.doOpen() as fileH:\n",
    "            \n",
    "            header = ''\n",
    "            sequence = ''\n",
    "            \n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>') :\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            for line in fileH:\n",
    "                if line.startswith ('>'):\n",
    "                    yield header,sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else :\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header,sequence\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Main \n",
    "Here is a jupyter framework that may come in handy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "def main (fileName=None):\n",
    "    myReader = FastAreader(fileName) \n",
    "    myNuc = NucParams()\n",
    "    for head, seq in myReader.readFasta() :\n",
    "        myNuc.addSequence(seq)\n",
    "        \n",
    "    # sort codons in alpha order, by Amino Acid\n",
    "    \n",
    "    # calculate relative codon usage for each codon and print\n",
    "    for nucI in nucs:\n",
    "        ...\n",
    "        print ('{:s} : {:s} {:5.1f} ({:6d})'.format(nuc, aa, val*100, thisCodonComp[nuc]))\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main('testGenome.fa') # make sure to change this in order to use stdin\n",
    "    "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inspection Intro"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Composition dictionaries are initialized in NucParams class, and the method addSequence updates the dictionaries by iterating and adding nucleotides/AAs/codons with a ticker count of +1 ."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## sequenceAnalysis.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "# Name: Justin Jang (jjang12)\n",
    "# Group Members: Aster Lathbury(mlathbur), Kimberly Magpantay(klmagpan), Faiz Khan(faahkhan)\n",
    "\"\"\"\n",
    "Analyze a genome. R\n",
    "\n",
    "\n",
    "Methods: \n",
    "NucParams: Return counts of codons and their translated amino acids.\n",
    "ProteinParam: Return aa count, pI, molar/mass extinction, and molecular weight\n",
    "FastAreader: Read FastA files so we can take objects from the file to analyze in genomeAnalyzer\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "import sys\n",
    "class NucParams:\n",
    "    \"\"\"Track counts of codons and their translated amino acids.\n",
    "     nucleotide > codon\n",
    "     Methods:\n",
    "     _init_(self)\n",
    "     addSequence(self,sequence)\n",
    "     aaComposition(self)  (from NB)\n",
    "     nucComposition(self)  (from NB)\n",
    "     codonComposition(self)  (from NB)\n",
    "     nucCount(self)  (from NB)\n",
    "       \"\"\"\n",
    "    rnaCodonTable = {\n",
    "    # RNA codon table\n",
    "    # U  \n",
    "    'UUU': 'F', 'UCU': 'S', 'UAU': 'Y', 'UGU': 'C',  # UxU\n",
    "    'UUC': 'F', 'UCC': 'S', 'UAC': 'Y', 'UGC': 'C',  # UxC\n",
    "    'UUA': 'L', 'UCA': 'S', 'UAA': '-', 'UGA': '-',  # UxA\n",
    "    'UUG': 'L', 'UCG': 'S', 'UAG': '-', 'UGG': 'W',  # UxG\n",
    "    # C\n",
    "    'CUU': 'L', 'CCU': 'P', 'CAU': 'H', 'CGU': 'R',  # CxU\n",
    "    'CUC': 'L', 'CCC': 'P', 'CAC': 'H', 'CGC': 'R',  # CxC\n",
    "    'CUA': 'L', 'CCA': 'P', 'CAA': 'Q', 'CGA': 'R',  # CxA\n",
    "    'CUG': 'L', 'CCG': 'P', 'CAG': 'Q', 'CGG': 'R',  # CxG\n",
    "    # A\n",
    "    'AUU': 'I', 'ACU': 'T', 'AAU': 'N', 'AGU': 'S',  # AxU\n",
    "    'AUC': 'I', 'ACC': 'T', 'AAC': 'N', 'AGC': 'S',  # AxC\n",
    "    'AUA': 'I', 'ACA': 'T', 'AAA': 'K', 'AGA': 'R',  # AxA\n",
    "    'AUG': 'M', 'ACG': 'T', 'AAG': 'K', 'AGG': 'R',  # AxG\n",
    "    # G\n",
    "    'GUU': 'V', 'GCU': 'A', 'GAU': 'D', 'GGU': 'G',  # GxU\n",
    "    'GUC': 'V', 'GCC': 'A', 'GAC': 'D', 'GGC': 'G',  # GxC\n",
    "    'GUA': 'V', 'GCA': 'A', 'GAA': 'E', 'GGA': 'G',  # GxA\n",
    "    'GUG': 'V', 'GCG': 'A', 'GAG': 'E', 'GGG': 'G'  # GxG\n",
    "    }\n",
    "    dnaCodonTable = {key.replace('U','T'):value for key, value in rnaCodonTable.items()}\n",
    "\n",
    "    def __init__ (self, inString=''):\n",
    "        '''Set up dictionaries, set value to zero. Add a tally if sequence contains an amino acid.'''\n",
    "        \n",
    "        \n",
    "        #make dictionaries with keys of sequence composition, values counts set to 0\n",
    "        self.nucComp = {'G': 0,'U': 0,'A': 0,'C': 0,'T': 0,'N': 0}  # nucleotide composition\n",
    "        self.aaComp = {aa: 0 for aa in NucParams.rnaCodonTable.values()}  # amino acid composition, key is amino acid and value is set to 0\n",
    "        self.codonComp = {codon: 0 for codon in NucParams.rnaCodonTable.keys()} #codon sequence composition, key is codon and value is set to 0\n",
    "        self.addSequence(inString)  # call addSequence method with the inputted sequence\n",
    "               \n",
    "        \n",
    "    def addSequence (self, sequence):\n",
    "        '''Clean the input sequence and increment how many of each RNA codon and its amino acid symbols in a list.\n",
    "         Store in dictionaries self.codonComp and self.aaComp'''\n",
    "        \n",
    "        sequence = ''.join(sequence.split()).upper()  # remove whitespace, convert to uppercase\n",
    "\n",
    "        for base in self.nucComp:  # update nucleotide composition (nucComp) with # of each nucleotide in sequence\n",
    "            self.nucComp[base] += sequence.count(base)  #\n",
    "        \n",
    "       \n",
    "        cleanedCodons = []  # create empty list to hold 3-base codons\n",
    "        for i in range(0, len(sequence), 3):  # iterate over sequence inputted string in increments of 3 nucleotides\n",
    "            cleanedCodons.append(sequence[i:i+3])  # place at position 0, 3, 6, 9 ...\n",
    "        \n",
    "\n",
    "        for codonSeq in cleanedCodons:  # iterate over cleanedCodons[] list\n",
    "            codonSeq = codonSeq.replace('T','U')  # change from DNA to RNA format b/c we are matching sequence with codonComp which is RNA format\n",
    "            if 'N'in codonSeq:  # skip N\n",
    "                continue\n",
    "            if codonSeq in self.codonComp:  # compare sequence with codon composition dictionary and searches for a match\n",
    "                self.codonComp[codonSeq] += 1  # increment codon sequence value by 1 IF there is a match\n",
    "                \n",
    "                aa = self.rnaCodonTable[codonSeq] # \n",
    "                self.aaComp[aa] += 1 #increment amino acid value by 1\n",
    "        pass\n",
    "    def aaComposition(self):\n",
    "        '''Return a composition of amino acids'''\n",
    "        return self.aaComp\n",
    "    def nucComposition(self):\n",
    "        '''Return a compostion of nucleotides'''\n",
    "        return self.nucComp\n",
    "    def codonComposition(self):\n",
    "        '''Return a compostion of 3-letter codons'''\n",
    "        return self.codonComp\n",
    "    def nucCount(self):\n",
    "        '''Return a count of each kind of nucleotide'''\n",
    "        return sum(self.nucComp.values())\n",
    "\n",
    "class ProteinParam :\n",
    "    \"\"\"Calculate statistics of a given protein sequence. Take a protein string and calculate the physical-chemical properties of a protein sequence. \n",
    "Return:\n",
    "- the number of amino acids and total molecular weight,\n",
    "- Molar extinction coefficient and Mass extinction coefficient,\n",
    "- theoretical isoelectric point (pI)\n",
    "- amino acid composition\n",
    "\n",
    "\n",
    "# These tables are for calculating:\n",
    "#     molecular weight (aa2mw), along with the mol. weight of H2O (mwH2O)\n",
    "#     absorbance at 280 nm (aa2abs280)\n",
    "#     pKa of positively charged Amino Acids (aa2chargePos)\n",
    "#     pKa of negatively charged Amino acids (aa2chargeNeg)\n",
    "#     and the constants aaNterm and aaCterm for pKa of the respective termini\n",
    "#  Feel free to move these to appropriate methods as you like\n",
    "\n",
    "# As written, these are accessed as class attributes, for example:\n",
    "# ProteinParam.aa2mw['A'] or ProteinParam.mwH2O\n",
    "\n",
    "    #dictionary of amino acids and corresponding molecular weights\n",
    "    aa2mw = {\n",
    "        'A': 89.093,  'G': 75.067,  'M': 149.211, 'S': 105.093, 'C': 121.158,\n",
    "        'H': 155.155, 'N': 132.118, 'T': 119.119, 'D': 133.103, 'I': 131.173,\n",
    "        'P': 115.131, 'V': 117.146, 'E': 147.129, 'K': 146.188, 'Q': 146.145,\n",
    "        'W': 204.225,  'F': 165.189, 'L': 131.173, 'R': 174.201, 'Y': 181.189\n",
    "    }\n",
    "\n",
    "    #molecular weight water\n",
    "    mwH2O = 18.015\n",
    "    #AAs and their absorbance\n",
    "    aa2abs280 = {'Y':1490, 'W': 5500, 'C': 125}\n",
    "\n",
    "    #charged AAs and their pKa vals\n",
    "    aa2chargePos = {'K': 10.5, 'R':12.4, 'H':6}\n",
    "    aa2chargeNeg = {'D': 3.86, 'E': 4.25, 'C': 8.33, 'Y': 10}\n",
    "    #pKas of N- and C- terminus\n",
    "    aaNterm = 9.69\n",
    "    aaCterm = 2.34\n",
    "\n",
    "    def __init__ (self, proteinStr: str):\n",
    "        \"\"\"Initialize method for taking protein sequence as an inputted string\"\"\"\n",
    "\n",
    "        #empty list to hold individual AAs in protein sequence\n",
    "        splitAminos = []\n",
    "\n",
    "        #list allowedAminos is the AAs from aa2mw\n",
    "        allowedAminos = self.aa2mw.keys()\n",
    "\n",
    "        for char in proteinStr.upper():  # input to upper\n",
    "            #iterate thru each char in input str, \n",
    "            #if char in input, then append to splitAminos list.\n",
    "            if char in allowedAminos:\n",
    "                splitAminos.append(char)\n",
    "\n",
    "        #join strings in splitAminos, then separate them into individual AAs in the list.\n",
    "        list = ''.join(splitAminos)\n",
    "        # joins strings with no separator adding: ''\n",
    "        # splits joined 'protein' into a char array, assigns to l\n",
    "\n",
    "        ## Join the list of amino acids back into a single string, and convert to uppercase.\n",
    "        self.newProteinStr = ''.join(list).upper()\n",
    "        # initialize newProteinStr to joined l and makes uppercase\n",
    "        \n",
    "        # Create an empty dictionary to hold the amino acid composition of the protein.\n",
    "        self.aaComp = {}\n",
    "        \n",
    "        # creates an empty dictionary object to represent # of each amino Acid\n",
    "        for aminoAcid in self.aa2mw.keys():\n",
    "            # initializes aminoAcid, loops through aa2mw.keys\n",
    "            self.aaComp[aminoAcid] = self.newProteinStr.count(aminoAcid)\n",
    "            # assigns aminoAcid to the key, counts self.newProteinStr for the amino acid letter\n",
    "\n",
    "    def aaCount (self):\n",
    "        \"\"\"Return a single integer count of valid amino acid characters found.\"\"\"\n",
    "        return (len(self.newProteinStr))\n",
    "        # returns len of input self's newProteinStr object\n",
    "\n",
    "    def _charge_ (self,pH):\n",
    "        \"\"\" Calculate the net charge of a protein based on its amino acid composition and the pH of the surrounding environment. Use for pI method.\"\"\"   \n",
    "        #ask TA questions about _charge_\n",
    "        posCharge = sum(self.aaComp.get(aa, 0) * pow(10, ProteinParam.aa2chargePos.get(aa, 0)) / (pow(10, ProteinParam.aa2chargePos.get(aa, 0)) + pow(10, pH)) for aa in ['R', 'K', 'H'])  # calculate positive charge from equation\n",
    "        negCharge = sum(self.aaComp.get(aa, 0) * pow(10, pH) / (pow(10, ProteinParam.aa2chargeNeg.get(aa, 0)) + pow(10, pH)) for aa in ['D', 'E', 'C', 'Y'])  # calculate negative charge from equation\n",
    "        \n",
    "        nTermChrg = pow(10, ProteinParam.aaNterm) / (pow(10,ProteinParam.aaNterm) + pow(10,pH))\n",
    "        cTermChrg = pow(10,pH) / (pow(10,ProteinParam.aaCterm) + pow(10,pH))\n",
    "\n",
    "        netCharge = posCharge - negCharge + nTermChrg - cTermChrg\n",
    "        return netCharge\n",
    "\n",
    "    def pI(self):\n",
    "        \"\"\"Calculate the  particular pH that yields a neutral net charge\"\"\"\n",
    "        bestCharge = 100000000\n",
    "        bestpH = 0\n",
    "        for pH100 in range (0, 1400+1): # want this down to two precision points, so much put it from 1 to 1400 not 1 to 14\n",
    "            pH = (pH100)/100 #gets the two decimal numbers (gives actual pH)\n",
    "            currentCharge = self._charge_(pH) #runs pH through the charge method\n",
    "            keepCharge = abs(currentCharge) #call charge method, gets absolute value \n",
    "            \n",
    "            if keepCharge < bestCharge: # wants the closest value to 0, so keeps it if it is lower than previous\n",
    "                bestCharge = keepCharge #defines keepCharge as new lower value, runs it through the loop again \n",
    "                bestpH =  pH\n",
    "        return bestpH   \n",
    "        \n",
    "    \n",
    "\n",
    "    def aaComposition (self) :\n",
    "        \"\"\"return a dictionary keyed by single letter Amino acid code, and having associated values that are the counts of those amino acids in the sequence.\"\"\"\n",
    "        return self.aaComp #returns the dictionary setup at the beginning\n",
    "\n",
    "    def molarExtinction (self):\n",
    "        \"\"\"Return how much light a protein absorbs at a certain wavelength\"\"\"\n",
    "        return (self.newProteinStr.count('Y') * self.aa2abs280['Y']\n",
    "                     + self.newProteinStr.count('W') * self.aa2abs280['W']\n",
    "                     + self.newProteinStr.count('C') * self.aa2abs280['C'])\n",
    "        \n",
    "\n",
    "    def massExtinction (self):\n",
    "        \"\"\"\n",
    "        Calculate light absorbance divided by molecular weight of the input protein sequence.\n",
    "        \"\"\"\n",
    "        myMW =  self.molecularWeight()\n",
    "        return self.molarExtinction() / myMW if myMW else 0.0\n",
    "    \n",
    "\n",
    "    def molecularWeight (self):\n",
    "        \"\"\"Return molecular weight of the protein sequence.\"\"\"\n",
    "        waterWeight = self.mwH2O * (self.aaCount() - 1)\n",
    "        aaWeight = sum(count * self.aa2mw[aa] for aa, count in self.aaComp.items())\n",
    "        molecularWeight = aaWeight - waterWeight\n",
    "        return molecularWeight\n",
    "\n",
    "\n",
    "import sys\n",
    "class FastAreader :\n",
    "    ''' \n",
    "    Define objects to read FastA files.\n",
    "    \n",
    "    instantiation: \n",
    "    thisReader = FastAreader ('testTiny.fa')\n",
    "    usage:\n",
    "    for head, seq in thisReader.readFasta():\n",
    "        print (head,seq)\n",
    "    '''\n",
    "    def __init__ (self, fname=None):\n",
    "        '''contructor: saves attribute fname '''\n",
    "        self.fname = fname\n",
    "            \n",
    "    def doOpen (self):\n",
    "        ''' Handle file opens, allowing STDIN.'''\n",
    "        if self.fname is None:\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "        \n",
    "    def readFasta (self):\n",
    "        ''' Read an entire FastA record and return the sequence header/sequence'''\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        \n",
    "        with self.doOpen() as fileH:\n",
    "            \n",
    "            header = ''\n",
    "            sequence = ''\n",
    "            \n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>') :\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            for line in fileH:\n",
    "                if line.startswith ('>'):\n",
    "                    yield header,sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else :\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header,sequence\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## genomeAnalyzer.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence length = 2.21 Mb\n",
      "\n",
      "GC content = 55.7 %\n",
      "\n",
      "UAA : -   32.6% (  1041)\n",
      "UAG : -   38.6% (  1230)\n",
      "UGA : -   28.8% (   918)\n",
      "GCA : A   14.1% ( 10605)\n",
      "GCC : A   40.5% ( 30524)\n",
      "GCG : A   30.5% ( 22991)\n",
      "GCU : A   14.9% ( 11238)\n",
      "UGC : C   67.2% (  4653)\n",
      "UGU : C   32.8% (  2270)\n",
      "GAC : D   70.8% ( 22686)\n",
      "GAU : D   29.2% (  9372)\n",
      "GAA : E   23.0% ( 11602)\n",
      "GAG : E   77.0% ( 38951)\n",
      "UUC : F   60.4% ( 15880)\n",
      "UUU : F   39.6% ( 10404)\n",
      "GGA : G   13.5% (  7651)\n",
      "GGC : G   47.9% ( 27193)\n",
      "GGG : G   29.3% ( 16664)\n",
      "GGU : G    9.3% (  5294)\n",
      "CAC : H   79.6% (  9089)\n",
      "CAU : H   20.4% (  2334)\n",
      "AUA : I   45.1% ( 18134)\n",
      "AUC : I   30.1% ( 12096)\n",
      "AUU : I   24.8% (  9986)\n",
      "AAA : K   32.8% ( 12810)\n",
      "AAG : K   67.2% ( 26206)\n",
      "CUA : L   15.2% ( 11753)\n",
      "CUC : L   23.6% ( 18242)\n",
      "CUG : L   21.2% ( 16349)\n",
      "CUU : L   14.7% ( 11393)\n",
      "UUA : L    8.0% (  6190)\n",
      "UUG : L   17.3% ( 13335)\n",
      "AUG : M  100.0% ( 13577)\n",
      "AAC : N   74.6% ( 12962)\n",
      "AAU : N   25.4% (  4423)\n",
      "CCA : P   16.4% (  6142)\n",
      "CCC : P   35.5% ( 13275)\n",
      "CCG : P   30.2% ( 11301)\n",
      "CCU : P   17.9% (  6690)\n",
      "CAA : Q   37.5% (  5738)\n",
      "CAG : Q   62.5% (  9544)\n",
      "AGA : R   21.5% ( 10672)\n",
      "AGG : R   46.1% ( 22871)\n",
      "CGA : R    2.8% (  1400)\n",
      "CGC : R   14.4% (  7139)\n",
      "CGG : R   10.8% (  5346)\n",
      "CGU : R    4.5% (  2237)\n",
      "AGC : S   25.5% (  9139)\n",
      "AGU : S    6.7% (  2383)\n",
      "UCA : S   10.6% (  3790)\n",
      "UCC : S   19.7% (  7062)\n",
      "UCG : S   20.1% (  7189)\n",
      "UCU : S   17.5% (  6253)\n",
      "ACA : T   22.7% (  7308)\n",
      "ACC : T   28.7% (  9253)\n",
      "ACG : T   29.9% (  9655)\n",
      "ACU : T   18.7% (  6029)\n",
      "GUA : V   18.2% ( 13288)\n",
      "GUC : V   22.8% ( 16617)\n",
      "GUG : V   43.0% ( 31384)\n",
      "GUU : V   16.0% ( 11698)\n",
      "UGG : W  100.0% ( 10915)\n",
      "UAC : Y   73.6% ( 22872)\n",
      "UAU : Y   26.4% (  8196)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<__main__.GenomeAnalyser at 0x15d23693710>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#!/usr/bin/env python3\n",
    "# Name: Justin Jang (jjang12)\n",
    "# Group Members: Aster Lathbury(mlathbur), Kimberly Magpantay(klmagpan), Faiz Khan(faahkhan)\n",
    "\"\"\"Analyze a genome from a FastaAfile.\"\"\"\n",
    "\n",
    "from sequenceAnalysis import NucParams, ProteinParam, FastAreader\n",
    "\n",
    "class GenomeAnalyser:\n",
    "    '''Display data from a Fasta file, in this case \"testGenome.fa\". Print sequence length, G-C content, and output statistics on \n",
    "    relative codon usage for each codon ordered by codon within each amino acid group. '''\n",
    "\n",
    "    def __init__(self,fileName = 'testGenome.fa'):\n",
    "        '''Constructor for GenomeAnalyser class.Take file and run the analyser method on it.'''\n",
    "        #useful dictionaries might use?\n",
    "        #add more stuff if needed\n",
    "        self.fastAreader = FastAreader()  \n",
    "        self.nucParams = NucParams()\n",
    "        \n",
    "        if fileName == None:  #in case fileName is none\n",
    "            fileName = sys.stdin\n",
    "        else:\n",
    "            self.analyser(fileName)\n",
    "        \n",
    "    \n",
    "    \n",
    "    def sequenceLength(self,read):\n",
    "        '''Run NucParams on the fastA file and return megabases'''\n",
    "        nucParams = NucParams()  # easier way to call NucParams methods\n",
    "        for head,seq in read.readFasta():  # note to self: takes readFasta from FastAreader in sequenceAnalysis\n",
    "            nucParams.addSequence(seq)\n",
    "\n",
    "        length = nucParams.nucCount() / 1000000  # return megabases by dividing by 1,000,000\n",
    "        return length, nucParams\n",
    "    \n",
    "    def gcContent(self, nucParams):\n",
    "        '''Calculate the percentage of G and C nucleotides in sequence from the file.'''\n",
    "        nucComp = nucParams.nucComposition() #see NucParams in sequenceAnalysis\n",
    "        gc = nucComp['G'] + nucComp['C']\n",
    "        gc = gc / nucParams.nucCount() * 100  #see NucParams in sequenceAnalysis\n",
    "\n",
    "        return gc\n",
    "\n",
    "    def analyser(self, fastAfile):\n",
    "        self.read = FastAreader(fastAfile)\n",
    "\n",
    "       \n",
    "        len, nucParams = self.sequenceLength(self.read)  # call sequenceLength with self.read as the argument, return len = length of sequence, nucParams = instance of NucParams with compositions stats\n",
    "        print(f\"sequence length = {len:.2f} Mb\\n\")\n",
    "\n",
    "        \n",
    "        self.GC = self.gcContent(nucParams)  # call gcContent with nucParams as the argument, which assigns GC content to the variable self.GC\n",
    "        print(f\"GC content = {self.GC:.1f} %\\n\")\n",
    "\n",
    "        codonCount = nucParams.codonComposition() # store codon composition stats in codonCount\n",
    "        \n",
    "\n",
    "        for codon in sorted(codonCount.keys(), key=lambda c: (nucParams.rnaCodonTable[c], c)):  # csort alphabetically first by 1-letter amino acid code, THEN by 3-letter amino acid code.\n",
    "        #is this a troll way of commenting idk                                                                                        # c represents each codon in codonCount\n",
    "            cCount = codonCount[codon] # count of current codon, store in cCount\n",
    "            singLetter = nucParams.rnaCodonTable[codon] # store single-letter AA in singLetter\n",
    "            aminoCount = nucParams.aaComp[singLetter] # store count of AA in aminoCount\n",
    "\n",
    "            # calculate frequencies\n",
    "            if cCount > 0:\n",
    "                finalFreq = (cCount/aminoCount) *100\n",
    "                print(f'{codon} : {singLetter}  {finalFreq:5.1f}% ({cCount:6d})')\n",
    "            else:\n",
    "                finalFreq = (cCount) *100\n",
    "                print(f'{codon} : {singLetter}  {finalFreq:5.1f}% ({cCount:6d})')\n",
    "\n",
    "        #print(codonCount)  run this to test raw count before showing frequencies\n",
    "        #test for redundancy later\n",
    "    \n",
    "GenomeAnalyser()\n",
    "\n",
    "#compare results with test.out"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "nteract": {
     "transient": {
      "deleting": false
     }
    }
   },
   "source": [
    "## Inspection Results\n",
    "Who is your inspection team? What did they find? How did you decide to incorporate their suggestions?"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Faiz's Comments:** I'd recommend adding somewhere in your NucParams class a few lines that discard codons that contain N bases since the directions ask for it. As for what 'continue' does in my code, whenever the loop encounters a codon that contains the letter 'N', it skips that iteration of the loop and moves on to the next codon. Other than that your code is very organized and easy to read, good job!\n",
    "\n",
    "**Aster's Comments:** Don't forget to add a program overview for both of your sequenceAnalysis and genomeAnalysis modules. We modified our codon strings differently, but I think it is interesting how you separated the codons every 3 characters. I think making each purpose of your genomeAnalysis into a method makes a lot of sense. I personally put it all into the main function, but creating a class makes your code more readable. Also, don't forget the inspection intro (included before NucParams). Great job!"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Kimberly's Comments**: In your __init__ method of your NucParams class, the only difference between our code is how you created the aaComp dictionary. I instead manually created it, similar to your nucComp, and could've probably done it this easier way. In your addSequence method, the way you counted the nucleotides is different in comparison to mine as you counted the number of instances. I incremented each base by 1, similar to your coson and aa compositions. Also, in your genomeAnalysis, it's interesting how you separated your different functions into different methods. It makes it look very organized. I instead added everything to a single \"main()\" function. Lastly, your code is very easy to read through, especially with your detailed comments."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Response: I implemented a for loop that skips over Ns while iterating over the list of cleaned codons. Also added other overviews and stuff I missed."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
